Add recipe for ox-json

jlumpe opened the pull request:
### Brief summary of what the package does

JSON export back end for Org mode

### Direct link to the package repository

https://github.com/jlumpe/ox-json

### Your association with the package

Author

### Relevant communications with the upstream package maintainer

**None needed**

### Checklist

Please confirm with `x`:

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses).
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [ ] My elisp byte-compiles cleanly
- [x] `M-x checkdoc` is happy with my docstrings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [ ] I have confirmed some of these without doing them

Replacement for #6251, I ended up rewriting from scratch. There are a few byte compilation errors about unused arguments for functions that are required to have a particular signature.

alphapapa commented:
>  (To silence byte-compiler warnings about unused variables, just use a variable name that starts with an underscore.  The byte-compiler interprets this as an indication that this is a variable known not to be
used.)

alphapapa commented:
It's awesome that you implemented this as an Org Export backend!

* The command names seem ambiguous to me, and having `json` in the names twice is reduntant.  Instead of `org-json-export-as-json`, why not `org-json-export-to-buffer`?  And instead of `org-json-export-to-json`, why not `org-json-export-to-file`?

jlumpe commented:
For the command names, I actually just used `ox-html` as a template. If that's not the general pattern then I agree it's redundant.

It looks like it may be a good idea to submit it as one of the `contrib/` packages, but I'll probably wait and see if people find it useful here first.

jlumpe commented:
Renamed those functions, also added several new features. (Basic) automated tests now work on Emacs 25.

riscy commented:
This quick first pass is semi-automated - please forgive any oversights.

<kbd>M-x checkdoc</kbd> (using version 0.6.1):
<details><summary>Click to expand</summary>
```
Warning (emacs): 
ox-json.el:90: First line is not a complete sentence
Warning (emacs): 
ox-json.el:278: First line is not a complete sentence
Warning (emacs): 
ox-json.el:357: Argument `timestamp' should appear (as TIMESTAMP) in the doc string
Warning (emacs): 
ox-json.el:410: Argument `async' should appear (as ASYNC) in the doc string
Warning (emacs): 
ox-json.el:424: Argument `async' should appear (as ASYNC) in the doc string
Warning (emacs): 
ox-json.el:435: Argument `info' should appear (as INFO) in the doc string
Warning (emacs): 
ox-json.el:440: Argument `info' should appear (as INFO) in the doc string
Warning (emacs): 
ox-json.el:566: All variables and subroutines might as well have a documentation string
Warning (emacs): 
ox-json.el:629: Argument `type' should appear (as TYPE) in the doc string
Warning (emacs): 
ox-json.el:696: All variables and subroutines might as well have a documentation string
Warning (emacs): 
ox-json.el:831: Argument `info' should appear (as INFO) in the doc string
Warning (emacs): 
ox-json.el:831: Lisp symbol `top-level' should appear in quotes
Warning (emacs): 
ox-json.el:860: Argument `name' should appear (as NAME) in the doc string
Warning (emacs): 
ox-json.el:872: Documentation strings should not start or end with whitespace
Warning (emacs): 
ox-json.el:866: Some lines are over 80 columns wide
Warning (emacs): 
ox-json.el:910: Argument `link' should appear (as LINK) in the doc string
Warning (emacs): 
ox-json.el:950: All variables and subroutines might as well have a documentation string
Warning (emacs): 
ox-json.el:958: All variables and subroutines might as well have a documentation string
Warning (emacs): 
ox-json.el:471: Messages should start with a capital letter
Warning (emacs): 
ox-json.el:489: Messages should start with a capital letter
Warning (emacs): 
ox-json.el:501: Messages should start with a capital letter
Warning (emacs): 
ox-json.el:693: Messages should start with a capital letter
```
</details>

<kbd>M-x package-lint-current-buffer</kbd> (using version 20190707.1456):
<details><summary>Click to expand</summary>
```
49 issues found:

66:36: warning: This file is not in the `cl-lib' ELPA compatibility package: require `cl-lib' instead.
77:1: error: "org-json-default-type-exporters" doesn't start with package's prefix "ox-json".
96:1: error: "org-json-default-property-types" doesn't start with package's prefix "ox-json".
297:1: error: "org-json--merge-alists" doesn't start with package's prefix "ox-json".
309:1: error: "org-json--plists-get-default" doesn't start with package's prefix "ox-json".
317:1: error: "org-json--plists-get" doesn't start with package's prefix "ox-json".
321:1: error: "org-json--loop-plist" doesn't start with package's prefix "ox-json".
334:1: error: "org-json--plist-to-alist" doesn't start with package's prefix "ox-json".
342:1: error: "org-json-node-properties" doesn't start with package's prefix "ox-json".
347:1: error: "org-json--is-node" doesn't start with package's prefix "ox-json".
356:1: error: "org-json-timestamp-isoformat" doesn't start with package's prefix "ox-json".
408:1: error: "org-json-export-to-buffer" doesn't start with package's prefix "ox-json".
422:1: error: "org-json-export-to-file" doesn't start with package's prefix "ox-json".
434:1: error: "org-json--make-error-obj" doesn't start with package's prefix "ox-json".
439:1: error: "org-json--error" doesn't start with package's prefix "ox-json".
445:1: error: "org-json--type-error" doesn't start with package's prefix "ox-json".
459:1: error: "org-json-encode-bool" doesn't start with package's prefix "ox-json".
473:1: error: "org-json-encode-string" doesn't start with package's prefix "ox-json".
491:1: error: "org-json-encode-number" doesn't start with package's prefix "ox-json".
503:1: error: "org-json-encode-array-raw" doesn't start with package's prefix "ox-json".
516:1: error: "org-json-encode-alist-raw" doesn't start with package's prefix "ox-json".
530:1: error: "org-json-encode-plist-raw" doesn't start with package's prefix "ox-json".
537:1: error: "org-json-encode-auto" doesn't start with package's prefix "ox-json".
565:1: error: "org-json--get-type-encoder" doesn't start with package's prefix "ox-json".
570:1: error: "org-json-encode-with-type" doesn't start with package's prefix "ox-json".
584:1: error: "org-json-encode-array" doesn't start with package's prefix "ox-json".
599:1: error: "org-json-encode-alist" doesn't start with package's prefix "ox-json".
614:1: error: "org-json-encode-plist" doesn't start with package's prefix "ox-json".
628:1: error: "org-json-make-object" doesn't start with package's prefix "ox-json".
644:1: error: "org-json-export-data" doesn't start with package's prefix "ox-json".
654:1: error: "org-json-encode-tag-string" doesn't start with package's prefix "ox-json".
661:1: error: "org-json-export-secondary-string" doesn't start with package's prefix "ox-json".
681:1: error: "org-json-export-property-node" doesn't start with package's prefix "ox-json".
695:1: error: "org-json-export-timestamp-property" doesn't start with package's prefix "ox-json".
718:1: error: "org-json-export-contents" doesn't start with package's prefix "ox-json".
738:1: error: "org-json-get-property-type" doesn't start with package's prefix "ox-json".
762:1: error: "org-json-export-properties-alist" doesn't start with package's prefix "ox-json".
778:1: error: "org-json-export-node-base" doesn't start with package's prefix "ox-json".
812:1: error: "org-json-transcode-plain-text" doesn't start with package's prefix "ox-json".
821:1: error: "org-json-transcode-base" doesn't start with package's prefix "ox-json".
830:1: error: "org-json--get-doc-info-alist" doesn't start with package's prefix "ox-json".
843:1: error: "org-json-transcode-template" doesn't start with package's prefix "ox-json".
859:1: error: "org-json--is-drawer-property-name" doesn't start with package's prefix "ox-json".
865:1: error: "org-json--separate-drawer-properties" doesn't start with package's prefix "ox-json".
890:1: error: "org-json-transcode-headline" doesn't start with package's prefix "ox-json".
909:1: error: "org-json-link-properties" doesn't start with package's prefix "ox-json".
939:1: error: "org-json-transcode-link" doesn't start with package's prefix "ox-json".
949:1: error: "org-json-timestamp-properties" doesn't start with package's prefix "ox-json".
957:1: error: "org-json-transcode-timestamp" doesn't start with package's prefix "ox-json".
```
</details>

<kbd>M-x byte-compile-file</kbd> (using Emacs 26.2):
```
ox-json.el:859:1:Warning: Unused lexical argument `info'

In end of data:
ox-json.el:965:1:Warning: the function `json-mode' is not known to be defined.
```

<kbd>M-x check-declare-file</kbd> (optional -- not on the MELPA checklist):
No issues! 🎉

-------
- ox-json.el#L388: Use `nil` instead of `'()`
- In several places, such as ox-json.el#L94, parentheses shouldn't be by themselves on the line

purcell commented:
Thanks! Notes from me:
- `package-lint` reports issues here, so I think you need to run it again.
- Agree with @riscy re. the `(` and `)` on separate lines: please prefer to follow the community style, even if you don't enjoy.

jlumpe commented:
Ok, most of these fixed. For the package lint, all of the problems are with the variable/function names being prefixed with `org-json-` instead of `ox-json-`. Here I'm following the pattern of `ox-html`, `ox-latex`, and others.

There is one compile error, in the export-to-buffer function I try to activate JSON mode if `require 'json-mode` is successful:

```
    (when (require 'json-mode nil t)
      (with-current-buffer buffer
        (json-mode)))
```

The compiler complains about `json-mode` not being defined. Any tips for fixing that?

riscy commented:
> The compiler complains about json-mode not being defined. Any tips for fixing that?

You could instead add this to the top of your file:
```elisp
(require 'json-mode nil t)
(declare-function json-mode "ext:json.el" t)
```
and change your conditional to e.g. `(when (fboundp 'json-mode) ...)`

Meanwhile, trying <kbd>M-x checkdoc</kbd> (using version 0.6.1):
```
Warning (emacs): 
ox-json.el:441: Argument `info' should appear (as INFO) in the doc string
Warning (emacs): 
ox-json.el:446: Argument `info' should appear (as INFO) in the doc string
Warning (emacs): 
ox-json.el:572: All variables and subroutines might as well have a documentation string
Warning (emacs): 
ox-json.el:844: Argument `info' should appear (as INFO) in the doc string
Warning (emacs): 
ox-json.el:844: Lisp symbol `top-level' should appear in quotes
Warning (emacs): 
ox-json.el:872: Argument `name' should appear (as NAME) in the doc string
Warning (emacs): 
ox-json.el:878: Some lines are over 80 columns wide
Warning (emacs): 
ox-json.el:477: Messages should start with a capital letter
Warning (emacs): 
ox-json.el:495: Messages should start with a capital letter
Warning (emacs): 
ox-json.el:507: Messages should start with a capital letter
Warning (emacs): 
ox-json.el:705: Messages should start with a capital letter
```

The package-lint errors (re: _I'm following the pattern of ox-html, ox-latex_) can be fixed by disregarding this pattern, of course. :)  This will reduce the chances of a hypothetical package called `org-json` from stumbling into your namespace, which could save someone some grief.

jlumpe commented:
Cool, fixed the compiler error.

I don't understand the "Messages should start with a capital letter" warning, these are all generated for calls to a private function which generates an error message based on a string argument it gets, but this argument is inserted into a fixed format string which does start with a capital.

The remainder of the `checkdoc` warnings  are in private functions, can those be ignored?

I've also just gone ahead and changed all the prefixes to `ox-json-`.

riscy commented:
Very much appreciated regarding the prefixes -- I think being pedantic about this is good for the package ecosystem, so thanks.

> I don't understand the "Messages should start with a capital letter" warning

Oops, this looks like a small bug with checkdoc, which effectively searches for `*error` and then a string literal.  Please disregard!  (You could rename the function or reorder the arguments, but neither of those are necessary sine this is a bug.)

> The remainder of the `checkdoc` warnings are in private functions, can those be ignored?

I can't speak for anyone else merging, but typically not, just as a matter of review policy.  If it were to be relaxed, I'd have to start manually checking for (extra) exceptions when checkdoc doesn't pass, and ultimately these lints are a proxy for making the packages developer friendly as well as user friendly.

jlumpe commented:
Checkdoc errors fixed.

riscy commented:
Thanks for waiting, I've been in transit.  I took it for a test drive, and this looks great now.  Welcome to MELPA!

jlumpe commented:
Awesome, thank you for bearing with me while I got everything up to spec.
